From e71238ed5d324542e5acd0c738d33ede3d6a06a7 Mon Sep 17 00:00:00 2001
From: Alin Dobre <alin.m.dobre@intel.com>
Date: Tue, 4 Sep 2012 10:42:52 +0300
Subject: Resolve apkcerts error wrt bash argument list being too long

When the BUILD_NUMBER make variable contains a longer string (tested
with 69 bytes), the command line that is generated by the ABS is
longer than 128k, which is, at this moment, a bash maximum. This leads
to the following error:

make: execvp: /bin/bash: Argument list too long
make: *** [<product-out>/obj/PACKAGING/apkcerts_intermediates/<some-name>.txt] Error 127

The cause for this is that the command line contains all the
name="...", certificate="..." and private_key="..." strings echos,
along with the redirect file, multiplied by many times, as many as the
packages included in the compiled product.

The solution from this patch keeps only doing echos for the strict
information, and the rest of the echoing is done in a bash for loop.

Change-Id: I7c57b6c319cc1ed78e279479002b4801175405cb
Signed-off-by: Alin Dobre <alin.m.dobre@intel.com>
---
 core/Makefile |   19 ++++++++++---------
 1 files changed, 10 insertions(+), 9 deletions(-)

diff --git a/core/Makefile b/core/Makefile
index 5195208..acc73bf 100644
--- a/core/Makefile
+++ b/core/Makefile
@@ -267,15 +267,16 @@ $(APKCERTS_FILE):
 	@echo APK certs list: $@
 	@mkdir -p $(dir $@)
 	@rm -f $@
-	$(foreach p,$(PACKAGES),\
-	  $(if $(PACKAGES.$(p).EXTERNAL_KEY),\
-	    $(call _apkcerts_echo_with_newline,\
-	      'name="$(p).apk" certificate="EXTERNAL" \
-	      private_key=""' >> $@),\
-	    $(call _apkcerts_echo_with_newline,\
-	      'name="$(p).apk" certificate="$(PACKAGES.$(p).CERTIFICATE)" \
-	      private_key="$(PACKAGES.$(p).PRIVATE_KEY)"' >> $@)))
-	# In case value of PACKAGES is empty.
+	$(hide) for xline in \
+			$(foreach p,$(PACKAGES),\
+			$(if $(PACKAGES.$(p).EXTERNAL_KEY),\
+				'$(p) EXTERNAL',\
+				'$(p) $(PACKAGES.$(p).CERTIFICATE) $(PACKAGES.$(p).PRIVATE_KEY)')) ; \
+		do \
+			read pkg_name pkg_cer pkg_key <<<$${xline}; \
+			echo name=\"$${pkg_name}\" certificate=\"$${pkg_cer}\" private_key=\"$${pkg_key}\" >>$@ ; \
+		done
+	# In case $(PACKAGES) is empty.
 	$(hide) touch $@
 
 .PHONY: apkcerts-list
-- 
1.7.4.1

